### PR TITLE
feat(source): support ingesting protobuf map

### LINKED ISSUE #13387: connector: support protobuf map type in source
Prior to https://github.com/risingwavelabs/risingwave/issues/12126, a `map` field is accepted and map to `struct<key, value>[]` in RisingWave, which might not be expected as suggested by @neverchanje so disabled there.

`JSONB` might be another good candidate, which supports efficient indexing.

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
fix #13387

#17743

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18512?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#18512** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18512?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#18507** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18507?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by chenzl25 — APPROVED** [review]
LGTM

File: src/connector/codec/src/decoder/protobuf/parser.rs [hunk 1]
```diff
@@ -180,10 +180,43 @@ pub fn from_protobuf_value<'a>(
...
+
+            let mut key_builder = map_type.key().create_array_builder(map.len());
+            let mut value_builder = map_type.value().create_array_builder(map.len());
+            // NOTE: HashMap's iter order is non-deterministic, but MapValue's
+            // order matters. We sort by key here to have deterministic order
+            // in tests. We might consider removing this, or make all MapValue sorted
+            // in the future.
+            for (key, value) in map.iter().sorted_by_key(|(k, _v)| *k) {
+                key_builder.append(from_protobuf_value(
+                    field_desc,
```
Comment (xiangjinwu) [inline src/connector/codec/src/decoder/protobuf/parser.rs:208]: `map_desc.map_entry_key_field()`

File: src/connector/codec/src/decoder/protobuf/parser.rs [hunk 2]
```diff
@@ -180,10 +180,43 @@ pub fn from_protobuf_value<'a>(
...
+                    &key.clone().into(),
+                    map_type.key(),
+                )?);
+                value_builder.append(from_protobuf_value(field_desc, value, map_type.value())?);
```
Comment (xiangjinwu) [inline src/connector/codec/src/decoder/protobuf/parser.rs:212]: `map_desc.map_entry_value_field()`

The error was not detected as most branches does not check input parameter `field_desc`, with struct / list being the exceptions. Can we test with a map where its value type is a struct?

File: src/connector/codec/src/decoder/protobuf/parser.rs [hunk 3]
```diff
@@ -180,10 +180,43 @@ pub fn from_protobuf_value<'a>(
...
+                }
+            };
+
+            let DataType::Map(map_type) = type_expected else {
+                return Err(err());
+            };
+            let map_desc = kind.as_message().ok_or_else(err)?;
+            if !map_desc.is_map_entry() {
+                return Err(err());
+            }
```
Comment (xiangjinwu) [inline src/connector/codec/src/decoder/protobuf/parser.rs:198]: Why not `field_desc.is_map()`

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**